install/0000_00_cluster-version-operator_01_cvoconfig.crd.yaml

```diff
@@ -15,6 +15,9 @@ spec:
       storage: true
   # either Namespaced or Cluster
   scope: Namespaced
+  subresources:
```

Comment:
The larger implication here is a formalization of a more rigid hierarchy of "truth" for cluster configuration:

1. top level configuration (tectonic config -> ImageConfig, NetworkConfig, CloudConfig) representing cluster level truths in functional/component terms rather than operator terms
2. secondary input configurations like secrets or raw config, or lower level component configuration (the pull secret, the cloud provider secret, or the openshift-apiserver config objects)
3. the CVOConfig -> ClusterVersion object (discussed previously), which is the source of truth for the desired version state of the cluster
4. specific API objects for specific operators like ClusterIngress, all the network SDN config, network policy, etc

Those are the configuration inputs to a cluster.  There is a parallel hierarchy of operators:

1. Cluster Version Operator -> ensures all "Cluster Operators" (I like this name better in colloquial use the more I use it) are functional and up to date, based on Cluster Version
2. Second-Level-Operators -> Cluster Operator are responsible for materializing the config hierarchy at a cluster scope
3. All cluster operators are expected to report operatorstatus/operator.status, which is how an administrator / developer performs quick triage of the state of the cluster

By making this change we are implying that an operator is a controllable object that has state that is reconciled with the cluster, and that the controllable object is the common interface to "cluster operators".  I think that having a consistent set of control knobs for operators is the right thing to do (there are zero today, but the first one would be something like "paused"), rather than having to duck type control knobs on config (since the config hierarchy is not operator focused but functional areas focused).  

All "cluster operators" should report status.  It should be possible to list all cluster operators.  It should be possible to build a UI that summarizes the rough relationships between cluster operators (although in most cases that should not be necessary).

It should be the job of the CVO to summarize operator status underneath the state of an upgrade.  Ultimately this is the level of truth:

1. The CVO takes a ClusterVersion and ensures all cluster operators are up to date
2. The CVO summarizes the result of that operation and ensures ClusterVersion reports it
3. An admin or user interfaces with ClusterVersion to effect changes and see their outcome.  
4. A developer or support person looks at operator status.
5. If an operator fails to update, there should be a condition on ClusterVersion which indicates that updates are not progressing.
6. The CVO should be reporting metrics for consumption by telemetry for input to SRE about the state of the update - some of those metrics will be summarizations of the current state like "upgrade to 3.1.5 in progress".  Others will be "failing operator status".

Reply:
Since the CVO is a reconciler, I think it's appropriate to talk about CVOConfig/ClusterVersion having a spec "desired outcome" and a status "actual state".  The state the CVO might present to an end user via ClusterVersion would be different than the status passed to OperatorStatus/Operator.status - both are for humans, but one reflects the desired action in cluster version, and the other reflects the desired state the operators themselves are trying to achieve.

Reply:
Are you familiar with how OLM works? It would be much nicer if there was a united view of operators, and OLM already figured a lot of this out already (like having a UI 🙂).

> By making this change we are implying that an operator is a controllable object that has state that is reconciled with the cluster, and that the controllable object is the common interface to "cluster operators". 
Yes, it's called the "ClusterServiceVersion" in OLM.
The operators even report back the status of their components: https://github.com/operator-framework/operator-lifecycle-manager/blob/d99daea0086cf92976617dd9326d0ed2b7ba85cf/pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go#L2

I'm not saying OLM solved all of the pieces needed for bootstrapping, but quite a bit of the stuff being discussed here.

Reply:
Yes.  These are going to be a deliberate subset of what an OLM operator may support.  The conditions on CSV status are very relevant to CV.  The spec is not.

Reply:
The biggest difference is that CSV needs conditions, and it's possible that we'll want to drive a phase for CV or something similar to simplify user input (although I'm leaning towards not having a phase and instead displaying human readable status instead).